### PR TITLE
 count_words challenge.

### PR DESCRIPTION
**Pull request by GaiSamuel**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_count_words.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        self.assertEqual(
+            count_words("My name is Gai"), {"my": 1, "name": 1, "is": 1, "gai": 1}
+        )
+
+    def test_count_words_with_an_empty_string(self):
+        """count_words returns an empty dictionary for an empty string"""
+        self.assertEqual(count_words(""), {})
+
+    def test_count_words_with_punctuation(self):
+        """Throws an assertion error if the punctuation is declared in it's own string"""
```
Comment (Mohamed-Elnageeb) [inline solutions/tests/test_count_words.py:31]: I didn't understand this.

File: solutions/tests/test_count_words.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    def test_count_words_with_numbers(self):
+        """counts the number in the text as a word of it's own"""
+        self.assertEqual(
+            count_words("Gai is 24 years old"),
+            {"gai": 1, "is": 1, "24": 1, "years": 1, "old": 1},
+        )
+
+    def test_count_words_with_special_characters(self):
+        """counts words with special characters and the special characters as same word"""
```
Comment (Mohamed-Elnageeb) [inline solutions/tests/test_count_words.py:63]: for this does it mean that if the string was  : "Hello, Gai! How are you Gai?" 
it should count Gai! and Gai? as different strings ?

**Review by Mohamed-Elnageeb — COMMENTED** [review]
I reviewed your work Gai and it is super clean and efficient,Kudos to you ✨,
I only had a few comments on the unittests.

File: solutions/tests/test_count_words.py (hunk 1 above)
Comment (GaiSamuel) [reply solutions/tests/test_count_words.py:31]: If the text was Tom! and you enter "tom": 1, "!": 1 it gives an assertion error.

File: solutions/tests/test_count_words.py (hunk 2 above)
Comment (GaiSamuel) [reply solutions/tests/test_count_words.py:63]: I just tested the theory typed, and hello, Gai! is counted as one word, but if the comma is removed, then gai! and gai? are counted as different words. Now, to the above test, I was trying to see if the word and the special character are counted separately.

**Review by Mohamed-Elnageeb — DISMISSED** [review]
Good work!

File: solutions/tests/test_count_words.py (hunk 1 above)
Comment (GaiSamuel) [reply solutions/tests/test_count_words.py]: I also noticed I was wording the docstrings wrong. I just fixed everything.

**Review by Mohamed-Elnageeb — APPROVED** [review]
Looks good to me, well done Gai.

**Review by TekaMesfinAbel — APPROVED** [review]
The implementation of count_words is functional and includes a solid set of test cases that handle various edge cases and input types. The code demonstrates good practices, such as normalization of input text and appropriate error handling.
